File: services/solana-service-ts/src/utils/amount.ts
import Decimal from "decimal.js";

Decimal.set({ precision: 40, rounding: Decimal.ROUND_DOWN });

const MAX_U64 = new Decimal("18446744073709551615");

export function parseAmountToBaseUnits(amountStr: string, decimals: number): bigint {
  if (!amountStr || typeof amountStr !== "string") {
    throw new Error("Amount must be a non-empty string");
  }
  let amount: Decimal;
  try {
    amount = new Decimal(amountStr);
  } catch {
    throw new Error(`Invalid amount format: ${amountStr}`);
  }
  if (amount.isNegative() || amount.isZero()) {
    throw new Error("Amount must be positive");
  }
  if (!Number.isInteger(decimals) || decimals < 0 || decimals > 38) {
    throw new Error(`Invalid decimals: ${decimals}`);
  }
  const baseUnits = amount.times(new Decimal(10).pow(decimals));
  if (!baseUnits.isInteger()) {
    throw new Error(`Amount has more decimal places than token decimals (${decimals})`);
  }
  if (baseUnits.greaterThan(MAX_U64)) {
    throw new Error("Amount exceeds maximum representable value");
  }
  return BigInt(baseUnits.toFixed(0));
}

File: services/solana-service-rs/src/services/amount.rs
use rust_decimal::prelude::*;
use rust_decimal::Decimal;

use crate::error::AppError;

/// Converts a human-readable amount string (e.g. "1.5") to base units (e.g. 1_500_000 for USDC).
/// Uses rust_decimal to avoid f64 overflow and precision loss.
pub fn parse_amount_to_base_units(amount_str: &str, decimals: u8) -> Result<u64, AppError> {
    let amount = Decimal::from_str(amount_str)
        .map_err(|_| AppError::InvalidParams(format!("Invalid amount format: {amount_str}")))?;

    if amount.is_sign_negative() || amount.is_zero() {
        return Err(AppError::InvalidParams(
            "Amount must be positive".to_string(),
        ));
    }

    let multiplier = Decimal::from(10u64.pow(u32::from(decimals)));
    let base_units = amount * multiplier;

    base_units
        .to_u64()
        .ok_or_else(|| AppError::InvalidParams(format!("Amount {amount_str} exceeds maximum representable value")))
}
